### PR TITLE
prov/verbs: Fix fi_cq_sread and incorrect handling of disconnection event from peer

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
1) Add handling of infinite timeout in fi_cq_sread
   - The problem was reproduced in rdm_tagged_peek tests
2) Update handling of disconnection event to cleanup prepost send
queue
   - When there are some send requestes in postponed queue and
   remote side closes the connection between enpoints, this was a
   cause of requests' leak and triggers ASSERT in debug mode
   (during destroying of pool since some resources are busy)
3) Remove unused function

Change-Id: I8ca3c46d7fb5a57094e3c61a6ebf33df211f8112
Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

**Comment by gladkovdmitry17** [comment]
@a-ilango

File: prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c [hunk 1]
```diff
@@ -114,7 +112,7 @@ ssize_t fi_ibv_rdm_cq_sreadfrom(struct fid_cq *cq, void *buf, size_t count,
 				int timeout)
 {
 	size_t threshold = count;
-	uint64_t time_limit = fi_gettime_ms() + timeout;
+	uint64_t time_limit = ((timeout < 0) ? -1 : (fi_gettime_ms() + timeout));
```
Comment (a-ilango) [inline prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c]: Shouldn't time_limit be assigned 0 when timeout < 0? The check on line 138 is !time_limit for endless loop.

File: prov/verbs/src/ep_rdm/verbs_ep_rdm.c [hunk 2]
```diff
@@ -470,38 +470,6 @@ static int fi_ibv_rdm_ep_close(fid_t fid)
...
-		if (!flags || (flags & FI_READ)) {
-		}
-
-		if (!flags || (flags & FI_WRITE) || (flags & FI_WRITE)) {
-		}
-	}
-	return 0;
-}
-#endif /* 0 */
-
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_ep_rdm.c]: This change should be a separate commit.

File: prov/verbs/src/ep_rdm/verbs_rdm_cm.c [hunk 3]
```diff
@@ -620,16 +622,21 @@ fi_ibv_rdm_process_event_disconnected(struct fi_ibv_rdm_ep *ep,
...
 	VERBS_INFO(FI_LOG_AV,
 		   "Disconnected from conn %p, addr %s:%u\n",
 		   conn, inet_ntoa(conn->addr.sin_addr),
 		   ntohs(conn->addr.sin_port));
 
+	while (NULL !=
+		(request = fi_ibv_rdm_take_first_from_posted_queue(ep))) {
+		FI_IBV_RDM_DBG_REQUEST("to_pool: ", request, FI_LOG_DEBUG);
+		util_buf_release(fi_ibv_rdm_request_pool, request);
+	}
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: The disconnect handling changes should also be in a separate commit.

File: prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c [hunk 4]
```diff
@@ -135,8 +133,11 @@ ssize_t fi_ibv_rdm_cq_sreadfrom(struct fid_cq *cq, void *buf, size_t count,
 						    threshold - counter,
 						    src_addr);
 		counter += (ret > 0) ? ret : 0;
-	} while ((ret >= 0) && (counter < threshold ||
-		(timeout >= 0 && fi_gettime_ms() < time_limit)));
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c]: We can simplify this if we check for an infinite timeout (-1) up front, and set time_limit = SIZE_MAX above.  I think that's the intent of this patch.  This should allow:
```
while ((ret >= 0) && (counter < threshold) && (fi_gettime_ms() < time_limit))
```
Why is the logic OR between the counter threshold and time checks

File: prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c (hunk 1 above)
Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c]: Sure, you're right

File: prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c (hunk 4 above)
Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c]: Thank you! will be updated in proposed way

File: prov/verbs/src/ep_rdm/verbs_rdm_cm.c (hunk 3 above)
Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Okay)

File: prov/verbs/src/ep_rdm/verbs_ep_rdm.c (hunk 2 above)
Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_ep_rdm.c]: Okay :)

File: prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c (hunk 4 above)
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c]: @a-ilango - Does the existing while loop logic look correct to you?  It looks like the || should be &&, or I'm not understanding the check.

Comment (a-ilango) [inline prov/verbs/src/ep_rdm/verbs_cq_ep_rdm.c]: Yes. It should be &&. If timeout occurs, we need to stop reading the CQ even if counter hasn't reached the threshold.
The updated code looks correct.

File: prov/verbs/src/ep_rdm/verbs_rdm_cm.c [hunk 5]
```diff
@@ -615,21 +617,76 @@ ssize_t fi_ibv_rdm_conn_cleanup(struct fi_ibv_rdm_conn *conn)
 	return ret;
 }
 
+static int fi_ibv_rdm_poll_cq(struct fi_ibv_rdm_ep *ep)
```
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Seems like this function almost duplicates `fi_ibv_rdm_tagged_poll_send`. Is it possible to merge them (or refactor existing one) adding `util_buf_release(fi_ibv_rdm_request_pool, request);` instead of `assert(0)` in `wc_error` section to avoid code duplication?

Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Thank you! Avoidance of duplication sounds reasonable, but we don't need to process postponed send queue since it's impossible to send smth in case of peer issued disconnection. I guess, more suitable to wrap part of code from `wc_error` to separate function and also replace duplicated code by `fi_ibv_rdm_process_send_wc` function and newly added one. This is just to don't take away `static inline` definition of `fi_ibv_rdm_tagged_poll_send `and `fi_ibv_rdm_process_send_wc`.

Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Maybe call here common poll function instead this one and skip unneeded logic basing on endpoint or connection state? Doing like this, you may save inlining on hot path paying cost of couple checks. Anyway it's up to you ;)

Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Not sure that I got it correctly. Do you mean to stay all as is and add the additional check on connection state in `fi_ibv_rdm_tagged_poll_send` (where postponed send queue are emptied)?
Also seems when we are handling WorkComp `fi_ibv_rdm_tagged_poll_send` and we face with !WC_STATUS_OK, other WC won't be handled (only !WC_STATUS_OK will be handled).

Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: See `fi_ibv_rdm_tagged_poll`, it skips receive CQ polling if ep is closing. i think you may do the same for postponed processing.
BTW, all polling functions have out of date `tagged` prefix, it's used for all supported capabilities now.

Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Thanks, we'll remove this prefix to avoid name collision.

Seems, I got it correctly, I meant the same. Yes, `fi_ibv_rdm_tagged_poll `really skips CQ polling if ep was closed. But it doesn't look as really so what I need, since ep_closing flags is set when EP is closed. In my case I need to check `fi_ibv_rdm_conn::state`.

Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: yep

Comment (gladkovdmitry17) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Thanks for your feedback!
